### PR TITLE
update Appendix:Requirements section of DG

### LINKED ISSUE #245: Update Appendix:Requirements section of Developer Guide

### PR DESCRIPTION
**Pull request by BobbyZhouZijian**
fixes #245

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -389,144 +362,284 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
   Use case ends.
 
 <br>
 
-**Use case 10: Edit a task**
+**Use case: UC10 - Edit a lesson**
+
+**Guarantees**
+
+* A lesson will be edited only if the edited lesson is still valid.
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md]: Still valid?? awkward phrasing

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -543,15 +656,15 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
 
 3. A user with above average typing speed for regular English text (i.e. not code, not system admin commands) should be able to accomplish most of the tasks faster using commands than using a mouse.
 
 4. The product should be easy to use by a novice with no experience of using a task management application.
 
-5. The size of built-in module data should not exceed 10mb.
+5. The product should be lightweight and should not exceed 20mb in size.
 
-6. Documentation should be easy to read with proper highlighting.
+6. Documentation should be easy to read and user-centric with proper highlighting.
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md]: with proper callouts and markdown

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -560,7 +673,10 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
 ### Glossary
 
 * **Mainstream OS**: Windows, Linux, Unix, OS-X
-* **Task**: A task is a collection of details about a specific task that needs to be done, including description, title, tag, date and time.
+* **Task**: A task is a collection of details about a specific task that needs to be done. A task can either be an **event**
+or a **deadline**.
+* **lesson**: A lesson is a collection of details of a specific lesson which happens in a recurring
+manner with a specified start and end date and a start and end time.
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md:684]: what about module tag and optional description?

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -251,12 +251,12 @@ _{Explain here how the data archiving feature will be implemented}_
...
-* has a lot of projects
+* has a lot of projects and modules
 * can type fast
-* wants to track productivity
-* wants to track workload
+* wants to track productivity and workload
 
-**Value proposition**: Allows NUS students to manage their tasks and schedule faster than a typical mouse/GUI driven app
+**Value proposition**: Allows NUS students to manage their tasks and visualize 
+their workload and productivity.
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md]: manage their tasks and visualise it on a calendar as well as time spent on tasks and lessons

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -269,75 +269,44 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+| `* *  `  | user                    | find tasks by date                         | easily know what events or deadlines are on that date. |
 | `* * *`  | user                    | find tasks by title                        |                                     |
-| `* *  `  | user                    | find tasks by tag                          |                                     |
+| `* *  `  | user                    | find tasks by description                  |                                     |
+| `* *  `  | user                    | find tasks by module it is associated      |                                     |
+| `* *  `  | user                    | find completed and uncompleted deadlines   | have a quick glance of what are to be completed. |
 | `* * *`  | user                    | edit my tasks                              |                                     |
+| `* * *`  | user                    | maintain a list of lessons I currently have  | track  my lessons.                |
+| `* * *`  | user with a lot of tasks| delete my lesson after it is not relevant   | focus on only the lessons I currently take. |
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md]: user with a lot of lessons

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -389,144 +363,284 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
 <br>
 
-**Use case 06: List all tasks**
+**Use case: UC05 - Find tasks**
 
 **MSS**
 
-1. User requests to list tasks.
+1. User <ins>lists a set of tasks (UC03 or UC05)</ins>
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md:427]: why did you quote UC05 isnt this UC05

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -389,144 +363,284 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
     Use case resumes at step 3.
 
 <br>
 
-**Use case 09: Find a task**
+**Use case: UC08 - Add a lesson**
+
+**Guarantees**
+
+* A new lesson will be added only there is no duplicate lesson already existing in PlaNus.
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md]: and does not overlap

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -389,144 +363,284 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+    Use case resumes at step 1.
+
+<br>
+
+**Use case: UC09 - Find lessons**
+
+**MSS**
+
+1. User <ins>lists a set of lessons (UC08 or UC09)</ins>
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md:549]: Why is it UC08/09

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -389,144 +363,284 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+**Use case: UC10 - Edit a lesson**
+
+**Guarantees**
+
+* A lesson will be edited only if the lesson is still valid after being edited.
 
 **MSS**
 
-1. User requests to edit values of specified attribute(s) of a task.
+1. User <ins>lists a set of lessons (UC08 or UC09)</ins>
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md:586]: ^

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -560,7 +674,13 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
 
 * **Mainstream OS**: Windows, Linux, Unix, OS-X
-* **Task**: A task is a collection of details about a specific task that needs to be done, including description, title, tag, date and time.
+* **Task**: A task is a collection of details about a specific task that needs to be done. A task can either be an **event**
+or a **deadline**.
+* **lesson**: A lesson is a collection of details of a specific lesson which happens in a recurring
+manner with a specified start and end date and a start and end time.
+* **tag**: A tag is an attribute that can be associated with a task or lesson to represent a module code. PlaNus' time analysis
+feature is based on the tags of the tasks and lessons.
+* **description**: A description is one or two sentences used to explain or add remarks to a task or lesson. It can be optional.
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md]: It is optional*

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -389,144 +363,284 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+  Use case ends.
+
 <br>
 
-**Use case 07: Request help**
+**Use case: UC06 - Edit a task**
+
+**Guarantees**
+
+* A task will be edited only if the task is still valid after being edited.
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md]: what does still valid mean? Maybe can specify that the task must not overlap with existing tasks or lessons after being edited

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -363,7 +333,11 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
     
 <br>
 
-**Use case 04: Add an event**
+**Use case: UC02 - Add an event**
+
+**Guarantees**
+
+* A deadline will be added only if there is no duplicate deadline existing in PlaNus.
```
Comment (MarcTzh) [inline docs/DeveloperGuide.md]: or overlaps with tasks and lessons

File: docs/DeveloperGuide.md (hunk 6 above)
Comment (BobbyZhouZijian) [reply docs/DeveloperGuide.md:427]: cuz u can list out a set of tasks using find as well?

File: docs/DeveloperGuide.md (hunk 8 above)
Comment (BobbyZhouZijian) [reply docs/DeveloperGuide.md:549]: Same. find can also list a set of lessons

**Review by MarcTzh — APPROVED** [review]
LGTM

**Pull request merged**